libr/core/cmd_debug.c

```diff
@@ -792,6 +795,192 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
+		ut64 bins = (NBYTES == 4) ? r_read_ble32 ((const void*) out, core->print->big_endian) : r_read_ble64 ((const void*) out, core->print->big_endian);
+		isNull = (bins == 0) ? true : false;
+		if (isNull) { 
+			eprintf (Color_BLUE"0x0 "Color_RESET""Color_GREEN"<repeats 254 times>"Color_RESET);
+			break;
+		} else eprintf (Color_BLUE" 0x%"PFMT64x""Color_RESET""Color_GREEN" <main_arena+%04d>, "Color_RESET, bins, bins-main_arena);
+	}
+
+	eprintf (Color_GREEN"\n\t}\t\n"Color_RESET);
+	eprintf (Color_GREEN"\tbinmap = {"Color_RESET);
+
+	offset = sizeof(size_t) * 266;
+	for(i = 0; i < 4; i++) {
+		r_core_read_at (core, main_arena + sizeof(int)*2 + offset + sizeof(int)*i, out, sizeof(int));
+		ut64 binmap = r_read_ble32 ((const void*) out, core->print->big_endian);
+		eprintf (Color_BLUE"0x%"PFMT64x""Color_RESET, binmap);
+		if (i < 3)
+			eprintf (Color_GREEN","Color_RESET);			
+	}
+	eprintf (Color_GREEN"}\n"Color_RESET);
+
+	const char *entry_str[] = {"next", "next_free", "system_mem", "max_system_mem"};
+
+	offset = (NBYTES == 4) ? sizeof(size_t) * 270 : sizeof(size_t) * 268;
+	for (i = 0; i < 4; i++) {
+		r_core_read_at (core, main_arena + sizeof(int)*2 + offset + sizeof(size_t)*i, out, sizeof(size_t));
+		ut64 entry = (NBYTES == 4) ? r_read_ble32 ((const void*) out, core->print->big_endian) : r_read_ble64 ((const void*) out, core->print->big_endian);
+		eprintf (Color_GREEN"\t%s = "Color_RESET""Color_BLUE" 0x%"PFMT64x""Color_RESET""Color_GREEN",\n"Color_RESET, entry_str[i], entry);
+	}
+	eprintf (Color_GREEN"}\n\n"Color_RESET);
+}
+
+static ut64 get_vaddr_symbol(const char *path, const char *symname) {
+	RListIter *iter;
+	RBinSymbol *s;
+	RCore *core = r_core_new ();
+	ut64 vaddr = 0LL;
+
+	r_bin_load (core->bin, path, 0, 0, 0, -1, false);
+	RList *syms = r_bin_get_symbols (core->bin);
+	r_list_foreach (syms, iter, s) {
+		if (strstr (s->name, symname)) {
+			vaddr = s->vaddr;		
+			break;
+		}
+	}	
+	r_core_free (core);
+
+	return vaddr;
+}
+
+static void get_hash_debug_directory(const char *path, char *hash) {
+	RListIter *iter;
+	RBinSection *s;
+	RCore *core = r_core_new ();
+	char buf[20] = {0};
+
+	r_bin_load (core->bin, path, 0, 0, 0, -1, false);
+	RList *sects = r_bin_get_sections (core->bin);
+	r_list_foreach (sects, iter, s) {
+		if (strstr (s->name, ".note.gnu.build-id")) {
+			r_io_read_at (core->io, s->vaddr+16, (ut8 *) buf, 20);
+			break;
+		}
+	}
+	r_core_free (core);
+
+	int i, j = 0;
+	for (i = 0; i < 20; i++) {
+		if (i <= 1) 
+			hash[i + 2 * j++] = (ut8) '/';
+
+		sprintf (hash + j + 2 * i, "%02x", (ut8) buf[i]);
+	}
+
+	strcat (hash, ".debug");
+}
+
+static int cmd_debug_map_heap(RCore *core, const char *input) {
+	const char* help_msg[] = {
+		"Usage:", "dmh", " # Memory map heap",
+		"dmha", "", "Struct Malloc State (main_arena)",
+		"dmh?", "", "",
+		NULL
+	};
+
+	RListIter *iter;
+	RDebugMap *map;
+	
+	switch (input[0]) {
+	
+	case 'a': // "dmha"
+		{
+		const char *dir_dbg = "/usr/lib/debug";
```

Comment:
Its ok for now to be hardcoded, but ideally configurable somehow later

> On 02 Jul 2016, at 04:03, Anton Kochkov <notifications@github.com> wrote:
> 
> In libr/core/cmd_debug.c:
> 
> > +static int cmd_debug_map_heap(RCore *core, const char *input) {
> > +	const char* help_msg[] = {
> > +		"Usage:", "dmh", " # Memory map heap",
> > +		"dmha", "", "Struct Malloc State (main_arena)",
> > +		"dmh?", "", "",
> > +		NULL
> > +	};
> > +
> > +	RListIter *iter;
> > +	RDebugMap *map;
> > +	
> > +	switch (input[0]) {
> > +	
> > +	case 'a': // "dmha"
> > +		{
> > +		const char *dir_dbg = "/usr/lib/debug";
> Those shouldn't be hardcoded, I think.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>